tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs

```diff
@@ -0,0 +1,343 @@
+        private static readonly Lazy<DebuggerManager> _lazyInstance =
+            new(
+                () => new DebuggerManager(DebuggerSettings.FromDefaultSource(), ExceptionReplaySettings.FromDefaultSource()),
+                LazyThreadSafetyMode.ExecutionAndPublication);
+
+        private readonly SemaphoreSlim _semaphore;
+        private readonly CancellationTokenSource _cancellationToken;
+        private volatile bool _isShuttingDown;
+        private int _initialized;
+
+        private DebuggerManager(DebuggerSettings debuggerSettings, ExceptionReplaySettings exceptionReplaySettings)
+        {
+            _initialized = 0;
+            _isShuttingDown = false;
+            _semaphore = new SemaphoreSlim(1, 1);
+            DebuggerSettings = debuggerSettings;
+            ExceptionReplaySettings = exceptionReplaySettings;
+            _cancellationToken = new CancellationTokenSource();
+
+            var tracerManager = TracerManager.Instance;
+            try
+            {
+                DynamicInstrumentationHelper.ServiceName = TraceUtil.NormalizeTag(tracerManager.Settings.ServiceName ?? tracerManager.DefaultServiceName);
+            }
+            catch (Exception e)
+            {
+                DynamicInstrumentationHelper.ServiceName = tracerManager.DefaultServiceName;
+                Log.Error(e, "Could not set `DynamicInstrumentationHelper.ServiceName`.");
+            }
+        }
+
+        internal static DebuggerManager Instance => _lazyInstance.Value;
+
+        internal DebuggerSettings DebuggerSettings { get; private set; }
+
+        internal ExceptionReplaySettings ExceptionReplaySettings { get; }
+
+        internal DynamicInstrumentation? DynamicInstrumentation { get; private set; }
+
+        internal SpanCodeOrigin.SpanCodeOrigin? CodeOrigin { get; private set; }
+
+        internal IDebuggerUploader? SymbolsUploader { get; private set; }
+
+        internal ExceptionDebugging? ExceptionReplay { get; private set; }
+
+        internal string ServiceName => DynamicInstrumentationHelper.ServiceName;
+
+        // /!\ This method is called by reflection in the Samples.SampleHelpers
+        // If you remove it then you need to provide an alternative way to wait for the discovery service
+        private static async Task<bool> WaitForDiscoveryService(IDiscoveryService discoveryService)
+        {
+            var tc = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            // Stop waiting if we're shutting down
+            LifetimeManager.Instance.AddShutdownTask(_ => tc.TrySetResult(false));
+
+            discoveryService.SubscribeToChanges(Callback);
+            return await tc.Task.ConfigureAwait(false);
+
+            void Callback(AgentConfiguration x)
+            {
+                tc.TrySetResult(true);
+                discoveryService.RemoveSubscription(Callback);
+            }
```

Comment:
Do we really want to do that every time DI is (re)enabled?
We also call `AddShutdownTask` here that might result in a leakage, as well as subscribing to `discoveryService`. `WaitForDiscoveryService` result might be cached and reused upon subsequent re-enablement of DI.

Reply:
Fixed in 930289ec802b1f379a56a52f9d03ee6fd6d59509